src/components/GoodsList/GoodsList.tsx

```diff
@@ -0,0 +1,215 @@
+import React from 'react';
+
+enum SortBy {
+  name = 'name',
+  length = 'length',
+  default = '',
+}
+
+enum Length {
+  One = 1,
+  Two,
+  Three,
+  Four,
+  Five,
+  Six,
+  Seven,
+  Eight,
+  Nine,
+  Ten,
+}
+
+type Props = {
+  goods: string[];
+};
+
+type State = {
+  isListVisible: boolean;
+  goods: string[];
+  isReversed: boolean;
+  sortBy: string;
+  visibleLength: number;
+};
+
+export class GoodsList extends React.Component<Props, State> {
+  defaultGoods = this.props.goods;
+
+  state: State = {
+    isListVisible: false,
+    goods: this.defaultGoods,
+    isReversed: false,
+    sortBy: SortBy.default,
+    visibleLength: Length.One,
+  };
+
+  toggleListVisibleHandler = () => {
+    this.setState(prevState => ({
+      isListVisible: !prevState.isListVisible,
+    }));
+  };
+
+  reverseHandler = () => {
+    this.setState(prevState => ({
+      isReversed: !prevState.isReversed,
+    }));
+  };
+
+  sortAlphabeticallyHandler = () => {
+    this.setState({
+      sortBy: SortBy.name,
+    });
+  };
+
+  resetHandler = () => {
+    this.setState({
+      goods: this.defaultGoods,
+      isReversed: false,
+      sortBy: SortBy.default,
+      visibleLength: Length.One,
+    });
+  };
+
+  sortByLengthHandler = () => {
+    this.setState({
+      sortBy: SortBy.length,
+    });
+  };
+
+  selectChangeHandler = (event: React.ChangeEvent<HTMLSelectElement>) => {
+    this.setState({
+      visibleLength: Number(event.target.value),
+    });
+  };
+
+  render() {
+    const {
+      isListVisible, goods, isReversed, sortBy, visibleLength,
+    } = this.state;
+
+    const copyGoods = [...goods].filter(item => item.length >= visibleLength);
+
+    copyGoods.sort((a, b) => {
+      switch (sortBy) {
+        case SortBy.name:
+          return a.localeCompare(b);
+
+        case SortBy.length:
+          return a.length - b.length;
+
+        default:
+          return 0;
+      }
+    });
+
+    if (isReversed) {
+      copyGoods.reverse();
+    }
```

Comment:
Btw, create a separate function for this logic. The method `render`, must be as concise as possible.

Reply:
Misha showed this a bit complicated solution to sort rendering array

Reply:
@Rozdorozhnii In the video, it's difficult to show a clear and optimized solution that would be clear to everyone, so he does not emphasize this

Reply:
So put this logic in a separate function